staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go

```diff
@@ -512,6 +512,62 @@ func PollImmediateUntil(interval time.Duration, condition ConditionFunc, stopCh
 	}
 }
 
+// PollUntilWithContext tries a condition func until it returns true,
+// an error or the specified context is cancelled or expired.
+//
+// PollUntilWithContext always waits interval before the first run of 'condition'.
+// 'condition' will always be invoked at least once.
+func PollUntilWithContext(interval time.Duration, condition ConditionFunc, ctx context.Context) error {
+	return WaitFor(poller(interval, 0), condition, ctx.Done())
+}
+
+// PollImmediateUntilWithContext tries a condition func until it returns true,
+// an error or the specified context is cancelled or expired.
+//
+// PollImmediateUntilWithContext runs the 'condition' before waiting for the interval.
+// 'condition' will always be invoked at least once.
+func PollImmediateUntilWithContext(interval time.Duration, condition ConditionFunc, ctx context.Context) error {
+	done, err := condition()
+	if err != nil {
+		return err
+	}
+	if done {
+		return nil
+	}
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+		return PollUntilWithContext(interval, condition, ctx)
+	}
+}
+
+// ContextFromChannelAndMaxWaitDuration returns a Context that is bound to the
+// specified channel and the wait duration. The derived context will be
+// cancelled when the specified channel stopCh is closed or the maximum wait
+// duration specified in maxWait elapses, whichever happens first.
+//
+// Note the caller must *always* call the CancelFunc, otherwise resources may be leaked.
+func ContextFromChannelAndMaxWaitDuration(stopCh <-chan struct{}, maxWait time.Duration) (context.Context, context.CancelFunc) {
```

Comment:
is this used outside of tests?  Why would something outside this package use this?

Reply:
I am planning to use it in a post start hook function for p&f, the use case is:
- A: post start hook provides us with a `StopCh` (may be in the future we will convert it to a context)
- B: and we want to try for `30s` at most

I need a derived context bound to both A and B that gets cancelled when either A or B elapses.

Reply:
How about fixing the post start hook functions to take a context?  There are fewer of those.

Reply:
@deads2k yes, we should convert the post start hook to accept `context`. but this is outside the scope of this PR. I agree that `ContextFromChannelAndMaxWaitDuration` does not naturally belong to `wait` package. So i removed it from this PR, I am going to define it where I am going to use it https://github.com/kubernetes/kubernetes/pull/98028